### PR TITLE
Private access credentials for private git repositories

### PR DESCRIPTION
**Pull request by frsyuki**
Pull-request #2825 added support for authorization credentials provided by bundle-config for rubygems source. This change adds the same support for HTTP(S) URL of git source.

Users can provide username and password as following:

    bundle config https://github.com/bundler/bundler.git username:password

or

    export BUNDLE_GITHUB__COM=username:password

Especially for private repositories on Github, users can use personal OAuth tokens:

    export BUNDLE_GITHUB__COM=abcd0123generatedtoken:x-oauth-basic

This enables us to easily deploy ruby code hosted on github private repositories to production servers where storing shared password in plain text at `~/.git-credentials` is security risk.

### CODE REVIEW

**Comment by yuuna** [comment]
:+1:

File: lib/bundler/source/git/git_proxy.rb [hunk 1]
```diff
@@ -136,15 +136,28 @@ def has_revision_cached?
...
             # Bash requires single quoted strings, with the single quotes escaped
             # by ending the string, escaping the quote, and restarting the string.
-            "'" + uri.gsub("'") { "'\\''" } + "'"
+            "'" + remote.gsub("'") {|s| "'\\''"} + "'"
+          end
+        end
+
+        # Adds credentials to the URI as Fetcher#configured_uri_for does
+        def configured_uri_for(uri)
+          if /https?:/ =~ uri
```
Comment (segiddins) [inline lib/bundler/source/git/git_proxy.rb]: why not just use `URI.regexp` here?

**Comment by segiddins** [comment]
Looks pretty good! Any way you could add some tests?

File: lib/bundler/source/git/git_proxy.rb (hunk 1 above)
Comment (frsyuki) [inline lib/bundler/source/git/git_proxy.rb]: I copied this code from `Fetcher#configured_uri_for`.
And more over, `URI.regexp =~ "git@github.com:bundler/bundler.git"` returns truthy but `URI("git@github.com:bundler/bundler.git")` raises exception:

```
irb(main):004:0> URI.regexp =~ "git@github.com:bundler/bundler.git"
=> 4
irb(main):006:0> URI.parse("git@github.com:bundler/bundler.git")
URI::InvalidURIError: bad URI(is not URI?): git@github.com:bundler/bundler.git
	from /Users/frsyuki/.rbenv/versions/2.2.2/lib/ruby/2.2.0/uri/rfc3986_parser.rb:66:in `split'
	from /Users/frsyuki/.rbenv/versions/2.2.2/lib/ruby/2.2.0/uri/rfc3986_parser.rb:72:in `parse'
	from /Users/frsyuki/.rbenv/versions/2.2.2/lib/ruby/2.2.0/uri/common.rb:226:in `parse'
	from (irb):6
```

Comment (segiddins) [inline lib/bundler/source/git/git_proxy.rb]: cool cool :+1:

**Comment by frsyuki** [comment]
I added a simple unit test. I was hard for me to actually test installation because faking private git repository is hard.

**Comment by segiddins** [comment]
:+1:

**Comment by indirect** [comment]
@homu r+

**Comment by seanarnold** [comment]
Just stumbled across this after a lot of searching, and it has solved my issue nicely. 
Is there any official documentation for this option? I can't see it here: https://bundler.io/v1.13/man/bundle-config.1.html @segiddins

**Comment by indirect** [comment]
@seanarnold good point. Opened a new ticket: https://github.com/bundler/bundler/issues/5172.

**Pull request merged**